File: src-tauri/src/commands/history.rs
use serde_json::Value;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use tauri::{AppHandle, Manager, Runtime, State};

use crate::core::database::Database;
use crate::core::history_store::{HistoryStore, HistoryStoreError};
use crate::core::paths::{PathKind, PathProvider};
use crate::integrations::asr::TranscriptSegment;
use crate::repositories::history::SqliteHistoryStore;
use crate::repositories::history::backup::{
    apply_prepared_history_import_inner, export_backup_archive_inner, prepare_backup_import_inner,
};
use crate::repositories::history::fs_utils::remove_path_if_exists;
use crate::repositories::history::{
    BackupManifest, ExportBackupArchiveRequest, HISTORY_DIR_NAME, HistoryAudioCleanupReport,
    HistoryAudioCleanupRequest, HistoryCreateLiveDraftRequest, HistoryItemRecord,
    HistoryListOptions, HistoryRepositoryState, HistorySaveImportedFileRequest,
    HistorySaveRecordingRequest, HistoryWorkspaceDateFilter, HistoryWorkspaceFilterType,
    HistoryWorkspaceQueryRequest, HistoryWorkspaceQueryResult, HistoryWorkspaceScope,
    HistoryWorkspaceSortOrder, LiveRecordingDraftResult, PreparedBackupImport,
    PreparedBackupImportState, TranscriptDiffResult, TranscriptDiffRow, TranscriptSnapshotMetadata,
    TranscriptSnapshotReason, TranscriptSnapshotRecord,
};

async fn run_history_file_task_inner<T, F>(
    app_local_data_dir: PathBuf,
    db: Arc<Database>,
    lock: Arc<Mutex<()>>,
    task: F,
) -> Result<T, String>
where
    T: Send + 'static,
    F: FnOnce(SqliteHistoryStore) -> Result<T, HistoryStoreError> + Send + 'static,
{
    tauri::async_runtime::spawn_blocking(move || {
        let _guard = lock.lock().map_err(|error| error.to_string())?;
        task(SqliteHistoryStore::new(app_local_data_dir, db)).map_err(|e| e.to_string())
    })
    .await
    .map_err(|error| error.to_string())?
}

async fn run_history_db_task<R, T, F>(app: &AppHandle<R>, task: F) -> Result<T, String>
where
    R: Runtime,
    T: Send + 'static,
    F: FnOnce(SqliteHistoryStore) -> Result<T, HistoryStoreError> + Send + 'static,
{
    let app_local_data_dir = (app as &dyn PathProvider).resolve_path(PathKind::AppLocalData)?;
    let db = Arc::clone(app.state::<Arc<Database>>().inner());
    crate::repositories::history::llm_helpers::run_llm_db_task(app_local_data_dir, db, task).await
}

async fn run_history_file_task<R, T, F>(
    app: &AppHandle<R>,
    state: State<'_, HistoryRepositoryState>,
    task: F,
) -> Result<T, String>
where
    R: Runtime,
    T: Send + 'static,
    F: FnOnce(SqliteHistoryStore) -> Result<T, HistoryStoreError> + Send + 'static,
{
    let app_local_data_dir = (app as &dyn PathProvider).resolve_path(PathKind::AppLocalData)?;
    let db = Arc::clone(app.state::<Arc<Database>>().inner());
    run_history_file_task_inner(app_local_data_dir, db, state.file_lock.clone(), task).await
}

#[tauri::command]
pub async fn history_list_items<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, HistoryRepositoryState>,
    limit: Option<usize>,
    offset: Option<usize>,
) -> Result<Vec<HistoryItemRecord>, String> {
    let opts = HistoryListOptions { limit, offset };
    run_history_file_task(&app, state, move |repository| {
        repository.list_items_with_reconciled_live_drafts_paginated(opts)
    })
    .await
}

#[tauri::command]
pub async fn history_query_workspace<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, HistoryRepositoryState>,
    scope: HistoryWorkspaceScope,
    query: String,
    filter_type: HistoryWorkspaceFilterType,
    date_filter: HistoryWorkspaceDateFilter,
    sort_order: HistoryWorkspaceSortOrder,
) -> Result<HistoryWorkspaceQueryResult, String> {
    let request = HistoryWorkspaceQueryRequest {
        scope,
        query,
        filter_type,
        date_filter,
        sort_order,
    };
    run_history_file_task(&app, state, move |repository| {
        repository.query_workspace(request)
    })
    .await
}

#[tauri::command]
pub async fn history_create_live_draft<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, HistoryRepositoryState>,
    id: Option<String>,
    audio_extension: String,
    project_id: Option<String>,
    icon: Option<String>,
) -> Result<LiveRecordingDraftResult, String> {
    let request = HistoryCreateLiveDraftRequest {
        id,
        audio_extension,
        project_id,
        icon,
    };
    run_history_file_task(&app, state, move |repository| {
        repository.create_live_draft(request)
    })
    .await
}

#[tauri::command]
pub async fn history_complete_live_draft<R: Runtime>(
    app: AppHandle<R>,
    history_id: String,
    segments: Value,
    duration: f64,
) -> Result<HistoryItemRecord, String> {
    run_history_db_task(&app, move |repository| {
        repository.complete_live_draft(&history_id, segments, duration)
    })
    .await
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn history_save_recording<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, HistoryRepositoryState>,
    segments: Value,
    duration: f64,
    project_id: Option<String>,
    audio_bytes: Option<Vec<u8>>,
    native_audio_path: Option<String>,
    audio_extension: Option<String>,
) -> Result<HistoryItemRecord, String> {
    let request = HistorySaveRecordingRequest {
        segments,
        duration,
        project_id,
        audio_bytes,
        native_audio_path,
        audio_extension,
    };
    run_history_file_task(&app, state, move |repository| {
        repository.save_recording(request)
    })
    .await
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn history_save_imported_file<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, HistoryRepositoryState>,
    id: Option<String>,
    source_path: String,
    segments: Value,
    duration: f64,
    project_id: Option<String>,
    converted_source_path: Option<String>,
) -> Result<HistoryItemRecord, String> {
    let request = HistorySaveImportedFileRequest {
        id,
        source_path,
        segments,
        duration,
        project_id,
        converted_source_path,
    };
    run_history_file_task(&app, state, move |repository| {
        repository.save_imported_file(request)
    })
    .await
}

#[tauri::command]
pub async fn history_delete_items<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, HistoryRepositoryState>,
    ids: Vec<String>,
) -> Result<(), String> {
    run_history_file_task(&app, state, move |repository| repository.delete_items(&ids)).await
}

#[tauri::command]
pub async fn history_load_transcript<R: Runtime>(
    app: AppHandle<R>,
    history_id: String,
) -> Result<Option<Vec<TranscriptSegment>>, String> {
    run_history_db_task(&app, move |repository| {
        repository.load_transcript(&history_id)
    })
    .await
}

#[tauri::command]
pub async fn history_update_transcript<R: Runtime>(
    app: AppHandle<R>,
    history_id: String,
    segments: Value,
) -> Result<HistoryItemRecord, String> {
    run_history_db_task(&app, move |repository| {
        repository.update_transcript(&history_id, segments)
    })
    .await
}

#[tauri::command]
pub async fn history_create_transcript_snapshot<R: Runtime>(
    app: AppHandle<R>,
    history_id: String,
    reason: TranscriptSnapshotReason,
    segments: Value,
) -> Result<TranscriptSnapshotMetadata, String> {
    run_history_db_task(&app, move |repository| {
        repository.create_transcript_snapshot(&history_id, reason, segments)
    })
    .await
}

#[tauri::command]
pub async fn history_list_transcript_snapshots<R: Runtime>(
    app: AppHandle<R>,
    history_id: String,
) -> Result<Vec<TranscriptSnapshotMetadata>, String> {
    run_history_db_task(&app, move |repository| {
        repository.list_transcript_snapshots(&history_id)
    })
    .await
}

#[tauri::command]
pub async fn history_load_transcript_snapshot<R: Runtime>(
    app: AppHandle<R>,
    history_id: String,
    snapshot_id: String,
) -> Result<Option<TranscriptSnapshotRecord>, String> {
    run_history_db_task(&app, move |repository| {
        repository.load_transcript_snapshot(&history_id, &snapshot_id)
    })
    .await
}

#[tauri::command]
pub fn history_build_transcript_diff(
    snapshot_segments: Vec<TranscriptSegment>,
    current_segments: Vec<TranscriptSegment>,
) -> Result<TranscriptDiffResult, String> {
    Ok(
        crate::repositories::history::transcript_diff::build_transcript_diff(
            snapshot_segments,
            current_segments,
        ),
    )
}

#[tauri::command]
pub fn history_restore_transcript_diff_rows(
    rows: Vec<TranscriptDiffRow>,
    selected_row_ids: Vec<String>,
) -> Result<Vec<TranscriptSegment>, String> {
    Ok(
        crate::repositories::history::transcript_diff::restore_transcript_diff_rows(
            rows,
            selected_row_ids,
        ),
    )
}

#[tauri::command]
pub async fn history_update_item_meta<R: Runtime>(
    app: AppHandle<R>,
    history_id: String,
    updates: Value,
) -> Result<(), String> {
    run_history_db_task(&app, move |repository| {
        repository.update_item_meta(&history_id, updates)
    })
    .await
}

#[tauri::command]
pub async fn history_update_project_assignments<R: Runtime>(
    app: AppHandle<R>,
    ids: Vec<String>,
    project_id: Option<String>,
) -> Result<(), String> {
    run_history_db_task(&app, move |repository| {
        repository.update_project_assignments(&ids, project_id)
    })
    .await
}

#[tauri::command]
pub async fn history_reassign_project<R: Runtime>(
    app: AppHandle<R>,
    current_project_id: String,
    next_project_id: Option<String>,
) -> Result<(), String> {
    run_history_db_task(&app, move |repository| {
        repository.reassign_project(current_project_id, next_project_id)
    })
    .await
}

#[tauri::command]
pub async fn history_load_summary<R: Runtime>(
    app: AppHandle<R>,
    history_id: String,
) -> Result<Option<Value>, String> {
    run_history_db_task(&app, move |repository| repository.load_summary(&history_id)).await
}

#[tauri::command]
pub async fn history_save_summary<R: Runtime>(
    app: AppHandle<R>,
    history_id: String,
    summary_payload: Value,
) -> Result<(), String> {
    run_history_db_task(&app, move |repository| {
        repository.save_summary(&history_id, summary_payload)
    })
    .await
}

#[tauri::command]
pub async fn history_delete_summary<R: Runtime>(
    app: AppHandle<R>,
    history_id: String,
) -> Result<(), String> {
    run_history_db_task(&app, move |repository| {
        repository.delete_summary(&history_id)
    })
    .await
}

#[tauri::command]
pub async fn history_resolve_audio_path<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, HistoryRepositoryState>,
    history_id: String,
) -> Result<Option<String>, String> {
    run_history_file_task(&app, state, move |repository| {
        repository.resolve_audio_path(&history_id)
    })
    .await
}

#[tauri::command]
pub async fn history_preview_audio_cleanup<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, HistoryRepositoryState>,
    retention_days: Option<u64>,
    exclude_history_id: Option<String>,
) -> Result<HistoryAudioCleanupReport, String> {
    let request = HistoryAudioCleanupRequest {
        retention_days,
        exclude_history_id,
    };
    run_history_file_task(&app, state, move |repository| {
        repository.preview_audio_cleanup(request)
    })
    .await
}

#[tauri::command]
pub async fn history_cleanup_audio<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, HistoryRepositoryState>,
    retention_days: Option<u64>,
    exclude_history_id: Option<String>,
) -> Result<HistoryAudioCleanupReport, String> {
    let request = HistoryAudioCleanupRequest {
        retention_days,
        exclude_history_id,
    };
    run_history_file_task(&app, state, move |repository| {
        repository.cleanup_audio(request)
    })
    .await
}

#[tauri::command]
pub async fn history_open_folder<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, HistoryRepositoryState>,
) -> Result<(), String> {
    let app_local_data_dir = (&app as &dyn PathProvider).resolve_path(PathKind::AppLocalData)?;
    let db = Arc::clone(app.state::<Arc<Database>>().inner());
    {
        let _guard = state.file_lock.lock().map_err(|error| error.to_string())?;
        SqliteHistoryStore::new(app_local_data_dir.clone(), db)
            .ensure_ready()
            .map_err(|e| e.to_string())?;
    }

    use tauri_plugin_opener::OpenerExt;
    app.opener()
        .open_path(
            app_local_data_dir.join(HISTORY_DIR_NAME).to_string_lossy(),
            None::<&str>,
        )
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn export_backup_archive<R: Runtime>(
    app: AppHandle<R>,
    history_state: State<'_, HistoryRepositoryState>,
    request: ExportBackupArchiveRequest,
) -> Result<BackupManifest, String> {
    let app_local_data_dir = (&app as &dyn PathProvider).resolve_path(PathKind::AppLocalData)?;
    let db = Arc::clone(app.state::<Arc<Database>>().inner());
    let lock = history_state.file_lock.clone();
    tauri::async_runtime::spawn_blocking(move || {
        let _guard = lock.lock().map_err(|error| error.to_string())?;
        export_backup_archive_inner(&app_local_data_dir, db, request)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn prepare_backup_import(
    state: State<'_, PreparedBackupImportState>,
    archive_path: String,
) -> Result<PreparedBackupImport, String> {
    let archive_path_buf = PathBuf::from(&archive_path);
    let (prepared, snapshot) = tauri::async_runtime::spawn_blocking(move || {
        prepare_backup_import_inner(&archive_path_buf)
    })
    .await
    .map_err(|error| error.to_string())??;

    state.insert(prepared.import_id.clone(), snapshot)?;
    Ok(prepared)
}

#[tauri::command]
pub async fn apply_prepared_history_import<R: Runtime>(
    app: AppHandle<R>,
    history_state: State<'_, HistoryRepositoryState>,
    prepared_state: State<'_, PreparedBackupImportState>,
    import_id: String,
) -> Result<(), String> {
    let Some(snapshot) = prepared_state.get(&import_id)? else {
        return Err(format!("Prepared backup import not found: {import_id}"));
    };

    let app_local_data_dir = (&app as &dyn PathProvider).resolve_path(PathKind::AppLocalData)?;
    let db = Arc::clone(app.state::<Arc<Database>>().inner());
    let lock = history_state.file_lock.clone();
    tauri::async_runtime::spawn_blocking(move || {
        let _guard = lock.lock().map_err(|error| error.to_string())?;
        apply_prepared_history_import_inner(
            &app_local_data_dir,
            db,
            &import_id,
            &snapshot.extraction_dir,
        )
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn dispose_prepared_backup_import(
    state: State<'_, PreparedBackupImportState>,
    import_id: String,
) -> Result<(), String> {
    let Some(snapshot) = state.remove(&import_id)? else {
        return Ok(());
    };

    tauri::async_runtime::spawn_blocking(move || {
        let _archive_path = snapshot.archive_path;
        remove_path_if_exists(&snapshot.extraction_dir)
    })
    .await
    .map_err(|error| error.to_string())?
}

File: platforms/desktop/frontend/src/services/tauri/history.ts
import {
  normalizeHistoryItemRecord,
  type HistoryAudioCleanupReport,
  type HistoryItem,
} from "../../types/history";
import type {
  HistoryItemMetaPatch_Serialize,
  HistoryItemRecord,
  HistoryWorkspaceQueryRequest as CoreHistoryWorkspaceQueryRequest,
  HistoryWorkspaceQueryResult as CoreHistoryWorkspaceQueryResult,
  HistoryWorkspaceScope,
  TranscriptDiffResult_Serialize,
  TranscriptDiffRow_Serialize,
  TranscriptSegment_Serialize,
  TranscriptSnapshotRecord_Serialize,
} from "../../bindings";
import type {
  HistorySummaryPayload,
  TranscriptSegment,
} from "../../types/transcript";
import type {
  TranscriptDiffResult,
  TranscriptDiffRow,
  TranscriptSnapshotMetadata,
  TranscriptSnapshotReason,
  TranscriptSnapshotRecord,
} from "../../types/transcriptSnapshot";
import {
  normalizeSpeakerAttribution,
  normalizeSpeakerTag,
} from "../../types/speaker";
import type { WorkspaceItemSearchMatch } from "../../utils/workspaceSearch";
import { TauriCommand } from "./commands";
import type { TauriCommandArgs, TauriCommandResult } from "./contracts";
import { invokeTauri } from "./invoke";

type HistoryDraftTransportHandle = TauriCommandResult<
  typeof TauriCommand.history.createLiveDraft
>;
type HistorySaveRecordingRequest = TauriCommandArgs<
  typeof TauriCommand.history.saveRecording
>;
type HistorySaveImportedFileRequest = TauriCommandArgs<
  typeof TauriCommand.history.saveImportedFile
>;
type HistoryAudioCleanupRequest = TauriCommandArgs<
  typeof TauriCommand.history.cleanupAudio
>;

function normalizeTranscriptSegment(segment: TranscriptSegment_Serialize): TranscriptSegment {
  return {
    id: segment.id,
    text: segment.text,
    start: segment.start,
    end: segment.end,
    isFinal: segment.isFinal,
    timing: segment.timing ?? undefined,
    tokens: segment.tokens ?? undefined,
    timestamps: segment.timestamps ?? undefined,
    durations: segment.durations ?? undefined,
    translation: segment.translation ?? undefined,
    speaker: normalizeSpeakerTag(segment.speaker) ?? undefined,
    speakerAttribution:
      normalizeSpeakerAttribution(segment.speakerAttribution) ?? undefined,
  };
}

function normalizeTranscriptDiffRow(row: TranscriptDiffRow_Serialize): TranscriptDiffRow {
  return {
    ...row,
    snapshotSegment: row.snapshotSegment
      ? normalizeTranscriptSegment(row.snapshotSegment)
      : undefined,
    currentSegment: row.currentSegment
      ? normalizeTranscriptSegment(row.currentSegment)
      : undefined,
  };
}

export interface HistoryDraftHandle<TItem = HistoryItemRecord> extends Omit<
  HistoryDraftTransportHandle,
  "item"
> {
  item: TItem;
}

export type HistoryWorkspaceQueryScope = HistoryWorkspaceScope;
export type HistoryWorkspaceQueryRequest = CoreHistoryWorkspaceQueryRequest;
export type HistoryWorkspaceQueryResult = Omit<
  CoreHistoryWorkspaceQueryResult,
  "filteredItems" | "searchMatchByItemId"
> & {
  filteredItems: HistoryItem[];
  searchMatchByItemId: Record<string, WorkspaceItemSearchMatch | null>;
};

export async function historyListItems(opts?: {
  limit?: number;
  offset?: number;
}): Promise<HistoryItemRecord[]> {
  return invokeTauri(TauriCommand.history.listItems, opts ?? {});
}

export async function historyCreateLiveDraft(
  id: string | null,
  audioExtension: string,
  projectId: string | null,
  icon: string | null,
): Promise<HistoryDraftHandle> {
  return invokeTauri(TauriCommand.history.createLiveDraft, {
    id,
    audioExtension,
    projectId,
    icon,
  });
}

export async function historyCompleteLiveDraft(
  historyId: string,
  segments: TranscriptSegment[],
  duration: number,
): Promise<HistoryItemRecord> {
  return invokeTauri(TauriCommand.history.completeLiveDraft, {
    historyId,
    segments,
    duration,
  });
}

export async function historySaveRecording(
  request: HistorySaveRecordingRequest,
): Promise<HistoryItemRecord> {
  return invokeTauri(TauriCommand.history.saveRecording, request);
}

export async function historySaveImportedFile(
  request: HistorySaveImportedFileRequest,
): Promise<HistoryItemRecord> {
  return invokeTauri(TauriCommand.history.saveImportedFile, request);
}

export async function historyDeleteItems(ids: string[]): Promise<void> {
  await invokeTauri(TauriCommand.history.deleteItems, { ids });
}

export async function historyLoadTranscript(
  historyId: string,
): Promise<TranscriptSegment[] | null> {
  const segments = await invokeTauri(TauriCommand.history.loadTranscript, { historyId });
  return segments?.map(normalizeTranscriptSegment) ?? null;
}

export async function historyUpdateTranscript(
  historyId: string,
  segments: TranscriptSegment[],
): Promise<HistoryItemRecord> {
  return invokeTauri(TauriCommand.history.updateTranscript, {
    historyId,
    segments,
  });
}

export async function historyCreateTranscriptSnapshot(
  historyId: string,
  reason: TranscriptSnapshotReason,
  segments: TranscriptSegment[],
): Promise<TranscriptSnapshotMetadata> {
  return invokeTauri(TauriCommand.history.createTranscriptSnapshot, {
    historyId,
    reason,
    segments,
  });
}

export async function historyListTranscriptSnapshots(
  historyId: string,
): Promise<TranscriptSnapshotMetadata[]> {
  return invokeTauri(TauriCommand.history.listTranscriptSnapshots, {
    historyId,
  });
}

export async function historyLoadTranscriptSnapshot(
  historyId: string,
  snapshotId: string,
): Promise<TranscriptSnapshotRecord | null> {
  const record: TranscriptSnapshotRecord_Serialize | null = await invokeTauri(
    TauriCommand.history.loadTranscriptSnapshot,
    { historyId, snapshotId },
  );
  return record
    ? { ...record, segments: record.segments.map(normalizeTranscriptSegment) }
    : null;
}

export async function historyBuildTranscriptDiff(
  snapshotSegments: TranscriptSegment[],
  currentSegments: TranscriptSegment[],
): Promise<{ rows: TranscriptDiffRow[]; changedCount: number }> {
  const result: TranscriptDiffResult_Serialize = await invokeTauri(
    TauriCommand.history.buildTranscriptDiff,
    { snapshotSegments, currentSegments },
  );
  return {
    ...result,
    rows: result.rows.map(normalizeTranscriptDiffRow),
  } satisfies TranscriptDiffResult;
}

export async function historyRestoreTranscriptDiffRows(
  rows: TranscriptDiffRow[],
  selectedRowIds: Iterable<string>,
): Promise<TranscriptSegment[]> {
  const segments = await invokeTauri(TauriCommand.history.restoreTranscriptDiffRows, {
    rows,
    selectedRowIds: Array.from(selectedRowIds),
  });
  return segments.map(normalizeTranscriptSegment);
}

export async function historyUpdateItemMeta(
  historyId: string,
  updates: HistoryItemMetaPatch_Serialize,
): Promise<void> {
  await invokeTauri(TauriCommand.history.updateItemMeta, {
    historyId,
    updates,
  });
}

export async function historyUpdateProjectAssignments(
  ids: string[],
  projectId: string | null,
): Promise<void> {
  await invokeTauri(TauriCommand.history.updateProjectAssignments, {
    ids,
    projectId,
  });
}

export async function historyReassignProject(
  currentProjectId: string,
  nextProjectId: string | null,
): Promise<void> {
  await invokeTauri(TauriCommand.history.reassignProject, {
    currentProjectId,
    nextProjectId,
  });
}

export async function historyLoadSummary(
  historyId: string,
): Promise<HistorySummaryPayload | null> {
  return invokeTauri(TauriCommand.history.loadSummary, { historyId });
}

export async function historySaveSummary(
  historyId: string,
  summaryPayload: HistorySummaryPayload,
): Promise<void> {
  await invokeTauri(TauriCommand.history.saveSummary, {
    historyId,
    summaryPayload,
  });
}

export async function historyDeleteSummary(historyId: string): Promise<void> {
  await invokeTauri(TauriCommand.history.deleteSummary, { historyId });
}

export async function historyResolveAudioPath(
  historyId: string,
): Promise<string | null> {
  return invokeTauri(TauriCommand.history.resolveAudioPath, { historyId });
}

export async function historyPreviewAudioCleanup(
  request: HistoryAudioCleanupRequest,
): Promise<HistoryAudioCleanupReport> {
  return invokeTauri(TauriCommand.history.previewAudioCleanup, request);
}

export async function historyCleanupAudio(
  request: HistoryAudioCleanupRequest,
): Promise<HistoryAudioCleanupReport> {
  return invokeTauri(TauriCommand.history.cleanupAudio, request);
}

export async function historyQueryWorkspace(
  request: HistoryWorkspaceQueryRequest,
): Promise<HistoryWorkspaceQueryResult> {
  const result = await invokeTauri(TauriCommand.history.queryWorkspace, request);
  const searchMatchByItemId = Object.fromEntries(
    Object.entries(result.searchMatchByItemId).map(([itemId, match]) => {
      if (!match) return [itemId, null];
      const matchedField =
        match.matchedField === "title" || match.matchedField === "previewText"
          ? match.matchedField
          : "searchContent";
      return [itemId, { ...match, matchedField } satisfies WorkspaceItemSearchMatch];
    }),
  );
  return {
    ...result,
    filteredItems: result.filteredItems.map(normalizeHistoryItemRecord),
    searchMatchByItemId,
  };
}

export async function historyOpenFolder(): Promise<void> {
  await invokeTauri(TauriCommand.history.openFolder);
}
